Sources/Foundation/JSONSerialization.swift

```diff
@@ -274,30 +339,37 @@ internal extension JSONSerialization {
         return .utf8
     }
     
-    static func parseBOM(_ bytes: UnsafePointer<UInt8>, length: Int) -> (encoding: String.Encoding, skipLength: Int)? {
-        if length >= 2 {
-            switch (bytes[0], bytes[1]) {
-            case (0xEF, 0xBB):
-                if length >= 3 && bytes[2] == 0xBF {
-                    return (.utf8, 3)
-                }
-            case (0x00, 0x00):
-                if length >= 4 && bytes[2] == 0xFE && bytes[3] == 0xFF {
-                    return (.utf32BigEndian, 4)
-                }
-            case (0xFF, 0xFE):
-                if length >= 4 && bytes[2] == 0 && bytes[3] == 0 {
-                    return (.utf32LittleEndian, 4)
-                }
-                return (.utf16LittleEndian, 2)
-            case (0xFE, 0xFF):
-                return (.utf16BigEndian, 2)
-            default:
-                break
-            }
+    static func parseBOM(_ bytes: UnsafeRawBufferPointer) -> (encoding: String.Encoding, skipLength: Int)? {
+        guard bytes.count >= 2 else {
+            return nil
+        }
+        
+        if bytes.starts(with: Self.utf8BOM) {
+            return (.utf8, 3)
+        }
+        if bytes.starts(with: Self.utf32BigEndianBOM) {
+            return (.utf32BigEndian, 4)
         }
+        if bytes.starts(with: Self.utf32LittleEndianBOM) {
+            return (.utf32LittleEndian, 4)
+        }
+        if bytes.starts(with: [0xFF, 0xFE]) {
+            return (.utf16LittleEndian, 2)
+        }
+        if bytes.starts(with: [0xFE, 0xFF]) {
+            return (.utf16BigEndian, 2)
+        }
+        
         return nil
```

Comment:
`detectEncoding` and `parseBOM` should be merged into one function returning `(encoding: String.Encoding, skipLength: Int)? ` since they both work on the same inputs.